clang/lib/CodeGen/CGCall.cpp

```diff
@@ -5322,6 +5324,11 @@ RValue CodeGenFunction::EmitCall(const CGFunctionInfo &CallInfo,
           IRCallArgs[FirstIRArg] = Val;
           break;
         }
+      } else if (I->getType()->isArrayParameterType()) {
+        // use the tmp created by the HLSLOutArgExpr
+        // instead of creating a new one below and copying the tmp into it.
+        IRCallArgs[FirstIRArg] = I->getKnownRValue().getScalarVal();
+        break;
```

Comment:
I think this might be the wrong solution. If we instead make the parameter ABI `Direct`, it should end up doing effectively what you're accomplishing here.

Reply:
Bleh... I spent some time today looking at this, and I think I've convinced myself that this is probably the simplest approach.

Without this you get a redundant copy because the cast that produces the ArrayParameterType. I have one small request to change the comment above.